connect_four.py

```diff
@@ -12,36 +17,34 @@ def __init__(self, c: int = 7, r: int = 6, w: int = 4) -> None:
         self.win_req = w #win requirement set by user
         self.current = [["_"]*self.num_cols for _ in range(self.num_rows)]
         self.columns = [[] for _ in range(c)]
+        self.board_full = False
 
     #string representation of the board 
     def __str__(self) -> str:
-
-    #function for printing out the current board and pieces played
-    def print_board(self) -> None: #TODO make this __str__ and add print output into game class
-
-        for row_count, row in enumerate(self.current):
-            print(f"{row_count}  {"  ".join(row)}")          
-
+        board_str = "\n".join(f"{row_count}  {'  '.join(row)}" for row_count, row in enumerate(self.current))
         column_title = [chr(x + ord("A")) for x in range(self.num_cols)] #create a list of column headers starting at "A"
-        print(f"   {"  ".join(column_title)}")
+        board_str += f"\n   {'  '.join(column_title)}"
 
+        return board_str
 
     #function for checking the validity of a chosen column. Checks if col exists and is not full. 
-    def check_col(self, col: str) -> int:   
-        col_i = ord(col) #using UNICODE of capital letters    
+    def check_col(self, col: str) -> ErrorCode:   
+        col_unicode = ord(col) #using UNICODE of capital letters    
         unicode_A = ord("A")  # ord("A") == 65 
-        if col_i < unicode_A or col_i > unicode_A -1 + self.num_cols: #max board size is 26 (A-Z)
-            return -1 #error code -1 = column doesn't exist. TODO replace with Enum codes
-        elif len(self.columns[col_i - unicode_A]) >= self.num_rows:
-            return -2 #error code -2 = column is full
+        if col_unicode < unicode_A or col_unicode > unicode_A -1 + self.num_cols: #max board size is 26 (A-Z)
+            return ErrorCode.COL_INVALID #error code 1 = column doesn't exist. 
+        elif len(self.columns[col_unicode - unicode_A]) >= self.num_rows:
+            return ErrorCode.COL_FULL #error code 2 = column is full
         else:
-            return col_i - unicode_A
+            return ErrorCode.VALID
 
-    def insert_new_piece(self, col: int, color: str) -> bool:
-        row = self.num_rows - len(self.columns[col])
+    def insert_new_piece(self, col: int, color) -> None:
+        row = (self.num_rows - 1 - len(self.columns[col]))
         self.columns[col].append(color)
         self.current[row][col] = color
-        return True
+        if sum(len(col) for col in self.columns) >= self.num_cols * self.num_rows:
```

Comment:
Instead of setting the state of board full, i think it's better to create a method for this that performs the same logic. So that would involve removing the logic here and creating a separate method that is called in cases where we're checking whether the board is full. The rationale is we want to minimize the amount of state in the program that could be unnecessary, and the value of self.board_full can be easily derived from the statement: 

if sum(len(col) for col in self.columns) >= self.num_cols * self.num_rows: